update(JS): web/javascript/guide/working_with_objects (#978)
webdoky/content: files/uk/web/javascript/guide/working_with_objects/index.md

  - Object
  - "l10n:priority"
---

{{jsSidebar("JavaScript Guide")}} {{PreviousNext("Web/JavaScript/Guide/Keyed_collections", "Web/JavaScript/Guide/Using_Classes")}}

JavaScript збудовано на основі простої, заснованої на об'єктах парадигми. Об'єкт — це колекція [властивостей](/uk/docs/Glossary/property/JavaScript), а властивість — це зв'язок між назвою (або іменем чи _ключем_) та значенням. Значенням властивості може бути функція — таку властивість називають [методом](/uk/docs/Glossary/Method).

Об'єкти в JavaScript, як і в багатьох інших мовах програмування, можна порівняти з об'єктами реального світу. В JavaScript об'єкт є окремою сутністю, що має властивості й тип. Порівняймо його, наприклад, з чашкою. Чашка є об'єктом, що має властивості. Чашка має колір, конструкцію, вагу, певний матеріал тощо. Так само об'єкти JavaScript можуть мати властивості, котрі визначають їх характеристики.

Додатково до об'єктів, уже визначених в браузері, можна створювати також власні об'єкти. Ця глава описує, як використовувати об'єкти, властивості, функції й методи, та як створювати власні об'єкти.

## Створення нових об'єктів

> Свої об'єкти можна створити, застосувавши [ініціалізатор об'єкта](/uk/docs/Web/JavaScript/Reference/Operators/Object_initializer). Окрім цього, можна також спершу створити функцію-конструктор, і потім реалізувати новий об'єкт шляхом звертання до функції з оператором `new`.

Author: ```suggestion
Свої об'єкти можна створити, застосувавши [об'єктний ініціалізатор](/uk/docs/Web/JavaScript/Reference/Operators/Object_initializer). Окрім цього, можна також спершу створити функцію-конструктор, і потім реалізувати новий об'єкт шляхом звертання до функції з оператором `new`.
```

> ### Застосування ініціалізаторів об'єкта

Reviewer: Є об'єктивні причини, чому це краще за попередній варіант?

Формлуювання "ініціалізатори об'єкта" натякає на те, що ініціалізаторів у об'єкта може бути декілька, чого б я краще уникав
```suggestion
### Застосування об'єктних ініціалізаторів
```

> Ініціалізатори об'єктів також звуть _літералами об'єктів_. Термін "об'єктний ініціалізатор" сумісний з термінологією, прийнятою в середовищі C++.

Author: ```suggestion
Об'єктні ініціалізатори також звуть _літералами об'єктів_. Термін "об'єктний ініціалізатор" сумісний з термінологією, прийнятою в середовищі C++.
```

Синтаксис створення об'єкта за допомогою об'єктного ініціалізатора має такий вигляд:

```js
const obj = {
  property1: value1, // назва властивості може бути ідентифікатором
  2: value2, // або числом
  "property n": value3, // чи рядком
};
```

> Кожна назва властивості до двокрапки — ідентифікатор (або ім'я, або число, або рядковий літерал), та кожний `valueN` — вираз, чиє значення присвоюється назві властивості. Ім'я властивості також може бути виразом; обчислювані ключі повинні бути огорнуті квадратними дужками. Довідкова сторінка [ініціалізаторів об'єкта](/uk/docs/Web/JavaScript/Reference/Operators/Object_initializer) містить детальніше пояснення синтаксису.

Author: ```suggestion
Кожна назва властивості до двокрапки — ідентифікатор (або ім'я, або число, або рядковий літерал), та кожний `valueN` — вираз, чиє значення присвоюється назві властивості. Ім'я властивості також може бути виразом; обчислювані ключі повинні бути огорнуті квадратними дужками. Довідкова сторінка [об'єктних ініціалізаторів](/uk/docs/Web/JavaScript/Reference/Operators/Object_initializer) містить детальніше пояснення синтаксису.
```

У цьому прикладі новостворений об'єкт присвоюється змінній `obj`, – це необов'язково. Якщо немає потреби звертатися до такого об'єкта деінде, не обов'язково присвоювати його змінній. (Зверніть увагу, що може бути потреба загорнути літерал об'єкта в дужки, якщо об'єкт зустрічається там, де очікується інструкція, щоб літерал не був сприйнятий за блокову інструкцію.)

> Ініціалізатори об'єктів – це вирази, і кожний такий ініціалізатор призводить до створення нового об'єкта щоразу, коли виконується інструкція, в якій він зустрічається. Ідентичні ініціалізатори об'єктів створюють окремі об'єкти, котрі при порівнянні не вважатимуться рівними одне одному.

Reviewer: хммм, є причини вважати, що `Ініціалізатор об'єкта` !== `об'єктний ініціалізатор`?

Author: ```suggestion
Об'єктні ініціалізатори – це вирази, і кожний такий ініціалізатор призводить до створення нового об'єкта щоразу, коли виконується інструкція, в якій він зустрічається. Ідентичні об'єктні ініціалізатори створюють окремі об'єкти, котрі при порівнянні не вважатимуться рівними одне одному.
```

```js
// Об'єкт тварини Animal, зі своїми властивостями та інкапсульованим методом
const Animal = {
  type: "Безхребетні", // Усталене значення властивості
  displayType() {
    // Метод, що виводитиме тип тварини
    console.log(this.type);
  },
};

// Створюється новий тип тварини під назвою animal1
const animal1 = Object.create(Animal);
animal1.displayType(); // Друкує: Безхребетні

// Створюється новий тип тварини під назвою fish
const fish = Object.create(Animal);
fish.type = "Риби";
fish.displayType(); // Друкує: Риби
```

## Об'єкти і властивості

Об'єкт JavaScript має пов'язані з ним властивості. Властивості об'єкта – по суті те саме, що й змінні, окрім того, що вони пов'язані з об'єктами, а не [областями видимості](/uk/docs/Glossary/Scope). Властивості об'єкта визначають його характеристики.

Наприклад, цей приклад створює об'єкт під назвою `myCar`, із властивостями під назвами `make`, `model` і `year`, чиї значення – `"Ford"`, `"Mustang"` і `1969`:

```js
const myCar = {
  make: "Ford",
  model: "Mustang",
  year: 1969,
};
```

> Подібно до змінних JavaScript, імена властивостей - чутливі до регістру. Вони можуть бути лише рядками чи Symbol: усі ключі [перетворюються на рядки](/uk/docs/Web/JavaScript/Reference/Global_Objects/String#zvedennia-do-riadka), окрім Symbol. [Індекси масивів](/uk/docs/Web/JavaScript/Reference/Global_Objects/Array#notes) фактично є властивостями з рядковими ключами, що містять цілі числа.

Reviewer: Це звучить неузгоджено. Я б з радістю перекладав назву типу (тільки писав би його так само з великої літери), але цю дискусію пропоную винести кудись
```suggestion
Подібно до змінних JavaScript, імена властивостей - чутливі до регістру. Вони можуть бути лише рядками чи примірниками Symbol: усі ключі [перетворюються на рядки](/uk/docs/Web/JavaScript/Reference/Global_Objects/String#zvedennia-do-riadka), окрім Symbol. [Індекси масивів](/uk/docs/Web/JavaScript/Reference/Global_Objects/Array#notes) фактично є властивостями з рядковими ключами, що містять цілі числа.
```

> Подібно до змінних JavaScript, імена властивостей - чутливі до регістру. Вони можуть бути лише рядками чи Symbol: усі ключі [перетворюються на рядки](/uk/docs/Web/JavaScript/Reference/Global_Objects/String#zvedennia-do-riadka), окрім Symbol. [Індекси масивів](/uk/docs/Web/JavaScript/Reference/Global_Objects/Array#indeksy-masyviv) фактично є властивостями з рядковими ключами, що містять цілі числа.

Author: ```suggestion
Подібно до змінних JavaScript, імена властивостей - чутливі до регістру. Вони можуть бути лише рядками чи примірниками Symbol: усі ключі [перетворюються на рядки](/uk/docs/Web/JavaScript/Reference/Global_Objects/String#zvedennia-do-riadka), окрім Symbol. [Індекси масивів](/uk/docs/Web/JavaScript/Reference/Global_Objects/Array#indeksy-masyviv) фактично є властивостями з рядковими ключами, що містять цілі числа.
```

  let result = "";
  for (const i in obj) {
    // Object.hasOwn() використано для виключення властивостей з
    // ланцюжка прототипів об'єкта і виведення винятково "власних властивостей"
    if (Object.hasOwn(obj, i)) {
      result += `${objName}.${i} = ${obj[i]}\n`;
    }
  }
  console.log(result);
}
```

Термін "власна властивість" позначає властивості об'єкта, але не включає властивості з ланцюжка прототипів. Тож виклик функції `showProps(myCar, 'myCar')` надрукує наступне:

```
myCar.make = Ford
myCar.model = Mustang
myCar.year = 1969
```

Код вище рівносильний наступному:

```js
function showProps(obj, objName) {
  let result = "";
  Object.keys(obj).forEach((i) => {
    result += `${objName}.${i} = ${obj[i]}\n`;
  });
  console.log(result);
}
```

Немає нативного способу перерахувати успадковані неперелічувані властивості. Проте цього можна досягти за допомогою наступної функції:

```js
function listAllProperties(myObj) {
  let objectToInspect = myObj;
  let result = [];
  while (objectToInspect !== null) {
    result = result.concat(Object.getOwnPropertyNames(objectToInspect));
    objectToInspect = Object.getPrototypeOf(objectToInspect);
  }
  return result;
}
```

> Докладніше про це – в [Перелічуваності та власництві властивостей](/uk/docs/Web/JavaScript/Enumerability_and_ownership_of_properties).

Reviewer: Тут вказується назва статті, я не певен, що такі речі варто відмінювати. Або якщо відмінювати - то писати з малої літери, щоб це не сприймалося як назва
```suggestion
Докладніше про це – у статті [Перелічуваність та власність властивостей](/uk/docs/Web/JavaScript/Enumerability_and_ownership_of_properties).
```

### Видалення властивостей

Неуспадковану властивість можна прибрати за допомогою оператора [`delete`](/uk/docs/Web/JavaScript/Reference/Operators/delete). Наступний код демонструє, як можна прибрати властивість.

```js
// Створюється новий об'єкт, myobj, з двома властивостями, a та b.
const myobj = new Object();
myobj.a = 5;
myobj.b = 12;
// Прибирається властивість a, залишаючи myobj лише властивість b.
delete myobj.a;
console.log("a" in myobj); // false
```

## Успадкування

Усі об'єкти в JavaScript успадковують принаймні від одного іншого об'єкта. Об'єкт, від якого успадковують, відомий як прототип, й успадковані властивості знаходяться в об'єкті у властивості конструктора `prototype`. Докладніше в [Успадкуванні й ланцюжку прототипів](/uk/docs/Web/JavaScript/Inheritance_and_the_prototype_chain).

### Означення властивостей для всіх об'єктів одного типу

> Додати властивість до всіх об'єктів, створених за допомогою одного [конструктора](#zastosuvannia-funktsii-konstruktora), можна за допомогою властивості [`prototype`](/uk/docs/Web/JavaScript/Reference/Global_Objects/Function/prototype). Такий підхід означає властивість, котру поділяють всі об'єкти конкретного типу, а не лише один примірник об'єкта. Наступний код додає властивість `color` до всіх об'єктів типу `Car`, а потім зчитує значення властивості з примірника `car1`.

Reviewer: Тут колізія, "означає" в сенсі "значити, позначати", і "означає" в сенсі "означити". Пропоную трохи додати слів для конкретизації
```suggestion
Додати властивість до всіх об'єктів, створених за допомогою одного [конструктора](#zastosuvannia-funktsii-konstruktora), можна за допомогою властивості [`prototype`](/uk/docs/Web/JavaScript/Reference/Global_Objects/Function/prototype). Такий підхід дає можливість означити властивість, котру поділяють всі об'єкти конкретного типу, а не лише один примірник об'єкта. Наступний код додає властивість `color` до всіх об'єктів типу `Car`, а потім зчитує значення властивості з примірника `car1`.
```

object.methodName(params);
```

> Методи здебільшого означають на об'єкті зі властивості конструктора `prototype`, аби усі об'єкти такого типу поділяли один метод. Наприклад, можна означити функцію, котра оформлює та виводить властивості раніше означених об'єктів `Car`.

Reviewer: ```suggestion
Означення методів здебільшого виконують на об'єкті зі властивості конструктора `prototype`, аби усі об'єкти такого типу поділяли один метод. Наприклад, можна означити функцію, котра оформлює та виводить властивості раніше означених об'єктів `Car`.
```

Intern.sayHi(); // Привіт, мене звати Олег
```

`this` – це "прихований параметр" виклику функції, котрий передається шляхом задання об'єкта перед функцією, котра викликається. Наприклад, для `Manager.sayHi()` `this` – це об'єкт `Manager`, тому що `Manager` стоїть перед функцією `sayHi()`. Якщо звернутись до тієї самої функції з іншого об'єкта, то `this` також зміниться. Якщо для виклику функції використовуються інші методи, як то {{jsxref("Function.prototype.call()")}} чи {{jsxref("Reflect.apply()")}}, то можна явно передати значення `this` у вигляді аргументу.

## Означення гетерів та сетерів

[Гетер] – це функція, пов'язана зі властивістю, котра отримує значення цієї конкретної властивості. [Сетер](/uk/docs/Web/JavaScript/Reference/Functions/set) – функція, пов'язана зі властивістю, котра задає значення цієї конкретної властивості. Вкупі гетер і сетер можуть опосередковано представляти значення властивості.

Гетери й сетери можуть бути:

- або означені всередині [об'єктних ініціалізаторів](#zastosuvannia-initsializatoriv-obiekta),
- або будь-коли пізніше додані до будь-якого наявного об'єкта.

> В [ініціалізаторах об'єктів](#zastosuvannia-initsializatoriv-obiekta) гетери й сетери означаються подібно до звичайних [методів](/uk/docs/Web/JavaScript/Reference/Functions/Method_definitions), але мають на початку означень ключові слова `get` або `set`. Метод-гетер не повинен очікувати на параметр, натомість метод-сетер очікує рівно на один параметр (нове значення, котре присвоюється). Наприклад:

Author: ```suggestion
В [об'єктних ініціалізаторах](#zastosuvannia-obiektnykh-initsializatoriv) гетери й сетери означаються подібно до звичайних [методів](/uk/docs/Web/JavaScript/Reference/Functions/Method_definitions), але мають на початку означень ключові слова `get` або `set`. Метод-гетер не повинен очікувати на параметр, натомість метод-сетер очікує рівно на один параметр (нове значення, котре присвоюється). Наприклад:
```
